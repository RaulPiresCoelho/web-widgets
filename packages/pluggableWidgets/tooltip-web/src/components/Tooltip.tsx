import { Placement } from "@floating-ui/react";
import classNames from "classnames";
import { createElement, CSSProperties, ReactElement, ReactNode, useState } from "react";
import { OpenOnEnum, RenderMethodEnum } from "../../typings/TooltipProps";
import { useFloatingUI } from "../utils/useFloatingUI";

export interface TooltipProps {
    name?: string;
    class?: string;
    style?: CSSProperties;
    tabIndex?: number;
    trigger: ReactNode;
    renderMethod: RenderMethodEnum;
    htmlMessage: ReactNode;
    textMessage?: string;
    position: Placement;
    openOn: OpenOnEnum;
    preview?: boolean;
}

export const Tooltip = (props: TooltipProps): ReactElement => {
    const { trigger, htmlMessage, textMessage, openOn, position, preview, renderMethod } = props;
    const [showTooltip, setShowTooltip] = useState(preview ?? false);
    const [arrowElement, setArrowElement] = useState<HTMLDivElement | null>(null);
<<<<<<< HEAD
    const { arrowStyles, blurFocusEvents, floatingStyles, getFloatingProps, getReferenceProps, refs, staticSide } =
=======
    const { refs, floatingStyles, staticSide, arrowStyles, getReferenceProps, getFloatingProps, blurFocusEvents } =
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        useFloatingUI({
            position,
            showTooltip,
            setShowTooltip,
            arrowElement,
            openOn
        });

    return (
        <div className={classNames(props.class, "widget-tooltip", `widget-tooltip-${position}`)}>
            <div
                className="widget-tooltip-trigger"
                ref={refs?.setReference}
<<<<<<< HEAD
                {...(preview
                    ? undefined
                    : getReferenceProps?.({ ...(openOn === "hoverFocus" && !preview ? blurFocusEvents : undefined) }))}
=======
                {...(preview ? undefined : getReferenceProps?.())}
                {...(openOn === "hoverFocus" && !preview ? blurFocusEvents : undefined)}
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            >
                {trigger}
            </div>
            {showTooltip && (textMessage || htmlMessage) ? (
                <div
                    className="widget-tooltip-content"
                    ref={refs?.setFloating}
                    style={floatingStyles}
                    {...getFloatingProps?.()}
                >
                    {renderMethod === "text" ? textMessage : htmlMessage}
                    <div className={`widget-tooltip-arrow-${staticSide}`} ref={setArrowElement} style={arrowStyles} />
                </div>
            ) : null}
        </div>
    );
};
