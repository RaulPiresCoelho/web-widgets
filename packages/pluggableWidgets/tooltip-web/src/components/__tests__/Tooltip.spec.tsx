import { createElement } from "react";
<<<<<<< HEAD
import { render, screen, act } from "@testing-library/react";
=======
import { render, screen } from "@testing-library/react";
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
import "@testing-library/jest-dom";

import { Tooltip, TooltipProps } from "../Tooltip";
import userEvent from "@testing-library/user-event";

<<<<<<< HEAD
jest.useFakeTimers();

=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
describe("Tooltip", () => {
    let defaultTooltipProps: TooltipProps;

    beforeEach(() => {
        defaultTooltipProps = {
            htmlMessage: undefined,
            textMessage: "Tooltip text",
            position: "right",
            openOn: "click",
            trigger: (
                <div tabIndex={0} data-testid="tooltip-trigger">
                    Trigger
                </div>
            ),
            renderMethod: "text",
            name: "tooltip"
        };
    });

    it("render DOM structure", async () => {
        const { asFragment } = render(<Tooltip {...defaultTooltipProps} />);
<<<<<<< HEAD
        const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
        const triggerElement = screen.getByTestId("tooltip-trigger");

        await user.click(triggerElement);
=======
        const triggerElement = screen.getByTestId("tooltip-trigger");

        await userEvent.click(triggerElement);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)

        expect(asFragment()).toMatchSnapshot();
    });

    it("opens tooltip onMouseEnter and close onMouseLeave", async () => {
        render(<Tooltip {...defaultTooltipProps} openOn="hover" />);
<<<<<<< HEAD
        const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
        const triggerElement = screen.getByTestId("tooltip-trigger");

        // hover
        await user.hover(triggerElement);
        act(() => {
            jest.advanceTimersByTime(100);
        });
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();

        // unhover
        await user.unhover(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeNull();

        // focus
        await user.keyboard("{Tab}");
=======
        const triggerElement = screen.getByTestId("tooltip-trigger");

        // hover
        await userEvent.hover(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();

        // unhover
        await userEvent.unhover(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeNull();

        // focus
        await userEvent.keyboard("{Tab}");
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        expect(triggerElement).toEqual(document.activeElement);
        expect(screen.queryByRole("tooltip")).toBeNull();

        // blur
<<<<<<< HEAD
        await user.keyboard("{Tab}");
=======
        await userEvent.keyboard("{Tab}");
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        expect(triggerElement).not.toEqual(document.activeElement);
        expect(screen.queryByRole("tooltip")).toBeNull();
    });

    it("open tooltip onClick", async () => {
        render(<Tooltip {...defaultTooltipProps} openOn="click" />);
<<<<<<< HEAD
        const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
        const triggerElement = screen.getByTestId("tooltip-trigger");
        // hover
        await user.hover(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeNull();

        // unhover
        await user.unhover(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeNull();

        // focus
        await user.keyboard("{Tab}");
=======
        const triggerElement = screen.getByTestId("tooltip-trigger");
        // hover
        await userEvent.hover(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeNull();

        // unhover
        await userEvent.unhover(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeNull();

        // focus
        await userEvent.keyboard("{Tab}");
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        expect(triggerElement).toEqual(document.activeElement);
        expect(screen.queryByRole("tooltip")).toBeNull();

        // blur
<<<<<<< HEAD
        await user.keyboard("{Tab}");
=======
        await userEvent.keyboard("{Tab}");
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        expect(triggerElement).not.toEqual(document.activeElement);
        expect(screen.queryByRole("tooltip")).toBeNull();

        // click
<<<<<<< HEAD
        await user.click(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();
        await user.click(triggerElement);
=======
        await userEvent.click(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();
        await userEvent.click(triggerElement);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        expect(screen.queryByRole("tooltip")).toBeNull();
    });

    it("open tooltip onFocus and close onBlur", async () => {
        render(<Tooltip {...defaultTooltipProps} openOn="hoverFocus" />);
<<<<<<< HEAD
        const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
        const triggerElement = screen.getByTestId("tooltip-trigger");

        // hover
        await user.hover(triggerElement);
        act(() => {
            jest.advanceTimersByTime(100);
        });
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();

        // unhover
        await user.unhover(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeNull();

        // focus
        await user.keyboard("{Tab}");
=======
        const triggerElement = screen.getByTestId("tooltip-trigger");

        // hover
        await userEvent.hover(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();

        // unhover
        await userEvent.unhover(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeNull();

        // focus
        await userEvent.keyboard("{Tab}");
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        expect(triggerElement).toEqual(document.activeElement);
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();

        // blur
<<<<<<< HEAD
        await user.keyboard("{Tab}");
=======
        await userEvent.keyboard("{Tab}");
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        expect(triggerElement).not.toEqual(document.activeElement);
        expect(screen.queryByRole("tooltip")).toBeNull();
    });

    it("render text content if the tooltipString is passed", async () => {
        render(<Tooltip {...defaultTooltipProps} renderMethod="text" />);
<<<<<<< HEAD
        const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
        const triggerElement = screen.getByTestId("tooltip-trigger");
        await user.click(triggerElement);
=======
        const triggerElement = screen.getByTestId("tooltip-trigger");
        await userEvent.click(triggerElement);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)

        expect(screen.queryByRole("tooltip")).toHaveTextContent(defaultTooltipProps.textMessage as string);
    });

    it("render HTML if the content is passed", async () => {
        render(
            <Tooltip
                {...defaultTooltipProps}
                renderMethod="custom"
                htmlMessage={<div>Simple Tooltip</div>}
                textMessage={undefined}
            />
        );
<<<<<<< HEAD
        const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

        const triggerElement = screen.getByTestId("tooltip-trigger");
        await user.click(triggerElement);
=======

        const triggerElement = screen.getByTestId("tooltip-trigger");
        await userEvent.click(triggerElement);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)

        expect(screen.queryByText("Simple Tooltip")).toBeInTheDocument();
    });

    it("close onOutsideClick if tooltip is visible", async () => {
        render(<Tooltip {...defaultTooltipProps} openOn="click" />);
<<<<<<< HEAD
        const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
        const triggerElement = screen.getByTestId("tooltip-trigger");
        await user.click(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();

        await user.click(document.body);
=======
        const triggerElement = screen.getByTestId("tooltip-trigger");
        await userEvent.click(triggerElement);
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();

        await userEvent.click(document.body);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        expect(screen.queryByRole("tooltip")).toBeNull();
    });
});
