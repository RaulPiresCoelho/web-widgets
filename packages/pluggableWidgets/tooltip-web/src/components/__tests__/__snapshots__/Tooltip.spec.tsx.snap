// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip render DOM structure 1`] = `
<DocumentFragment>
  <div
    class="widget-tooltip widget-tooltip-right"
  >
    <div
      aria-describedby=":r0:"
      class="widget-tooltip-trigger"
    >
      <div
        data-testid="tooltip-trigger"
        tabindex="0"
      >
        Trigger
      </div>
    </div>
    <div
      class="widget-tooltip-content"
<<<<<<< HEAD
      data-floating-ui-focusable=""
      id=":r0:"
      role="tooltip"
      style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px);"
=======
      id=":r0:"
      role="tooltip"
      style="position: absolute; left: 0px; top: 0px; transform: translate(8px, 0px);"
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
      tabindex="-1"
    >
      Tooltip text
      <div
        class="widget-tooltip-arrow-left"
<<<<<<< HEAD
=======
        style="top: 0px;"
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
      />
    </div>
  </div>
</DocumentFragment>
`;
